src/coreclr/jit/lowerarmarch.cpp

```diff
@@ -507,14 +506,50 @@ void Lowering::LowerCast(GenTree* tree)
 
     assert(!varTypeIsSmall(srcType));
 
-    if (tmpType != TYP_UNDEF)
+    // Transform '(ulong)smallInt' or '(uint)smallInt' to 'smallInt & 0xFF..'
+    if (comp->opts.OptimizationEnabled() && tree->gtGetOp1()->OperIs(GT_CAST) && !tree->gtOverflow() &&
+        tree->IsUnsigned() && !tree->isContained() && !varTypeIsSmallInt(dstType) &&
+        varTypeIsIntOrI(varTypeToSigned(dstType)))
     {
-        GenTree* tmp = comp->gtNewCastNode(tmpType, op1, tree->IsUnsigned(), tmpType);
-        tmp->gtFlags |= (tree->gtFlags & (GTF_OVERFLOW | GTF_EXCEPT));
+        GenTreeCast* cast2    = tree->gtGetOp1()->AsCast();
```

Comment:
Nit: `srcCast` would be more in line with other cast-related code (with `srcCastToType`).